pennylane/devices/qubit/simulate.py

```diff
@@ -415,71 +438,297 @@ def simulate_tree_mcm(
-    # update_mcm_samples and prune_mcm_samples respectively.
-    mcm_samples = mcm_samples or {}
-
-    circuit_base, circuit_next, op = circuit_up_to_first_mcm(circuit)
-    circuit_base = prepend_state_prep(circuit_base, interface, circuit.wires)
-    state, is_state_batched = get_final_state(
-        circuit_base,
-        debugger=debugger,
-        mid_measurements=mcm_active,
-        **execution_kwargs,
-    )
-    measurements = measure_final_state(circuit_base, state, is_state_batched, **execution_kwargs)
+    # `var` measurements cannot be aggregated on the fly as they require the global `expval`
+    # variance_transform replaces `var` measurements with `expval` and `expval**2` measurements
+    circuits, variance_post_processing = variance_transform(circuit)
+    circuit = circuits[0]
+    finite_shots = bool(circuit.shots)
+
+    ##################
+    # Parse MCM info #
+    ##################
+
+    # mcms is the list of all mid-circuit measurement operations
+    # mcms[d] is the parent MCM (node) of a circuit segment (edge) at depth `d`
+    # The first element is None because there is no parent MCM at depth 0
+    mcms = tuple([None] + [op for op in circuit.operations if isinstance(op, MidMeasureMP)])
+    n_mcms = len(mcms) - 1
+    # We obtain `measured_mcms_indices`, the list of MCMs which require post-processing:
+    # either as requested by terminal measurements or post-selection
+    measured_mcms = find_post_processed_mcms(circuit)
+    measured_mcms_indices = [i for i, mcm in enumerate(mcms[1:]) if mcm in measured_mcms]
+    # `mcm_samples` is a register of MCMs. It is necessary to correctly keep track of
+    # correlated MCM values which may be requested by terminal measurements.
+    mcm_samples = {
+        k + 1: qml.math.empty((circuit.shots.total_shots,), dtype=bool) if finite_shots else None
+        for k in measured_mcms_indices
+    }
+
+    #############################
+    # Initialize tree-traversal #
+    #############################
+
+    # mcm_current[:d+1] is the active branch at depth `d`
+    # The first entry is always 0 as the first edge does not stem from an MCM.
+    # For example, if `d = 2` and `mcm_current = [0, 1, 1, 0]` we are on the 11-branch,
+    # i.e. the first two MCMs had outcome 1. The last entry isn't meaningful until we are
+    # at depth `d=3`.
+    mcm_current = qml.math.zeros(n_mcms + 1, dtype=int)
+    # `mid_measurements` maps the elements of `mcm_current` to their respective MCMs
+    # This is used by `get_final_state::apply_operation` for `Conditional` operations
+    mid_measurements = dict(zip(mcms[1:], mcm_current[1:].tolist()))
+    # Split circuit into segments
+    circuits = split_circuit_at_mcms(circuit)
+    circuits[0] = prepend_state_prep(circuits[0], None, interface, circuit.wires)
+    terminal_measurements = circuits[-1].measurements if finite_shots else circuit.measurements
```

Comment:
Why would we have different measurements with finite shots then the analytic case?

Reply:
In the analytic case, MCM measurements are inserted in the measurement results during tree traversal and measurements are combined as is. In the finite shots case, MCM measurements are kept in an _MCM register_ (`mcm_samples`) which is used for post-processing (deals with post-selection and terminal measurements). It might be possible to use `insert_mcms` for finite shots too, but I think it will slow down `qml.sample` since it will trigger many concatenation operations.